From 98eb74839c7dfffd749b23c458335ea2c1a59c42 Mon Sep 17 00:00:00 2001
From: David Gow <david@ingeniumdigital.com>
Date: Sat, 8 Jan 2022 15:00:01 +0800
Subject: [PATCH 45/44] user32: Mask off the unspecified bits in GetKeyboardState()

When using shared memory for the keyboard state (as added in d2cbafab08)
we no-longer mask off the unused bits, which causes the map to scroll
uncontrollably in some situations in the Age of Empires series.

This is a regression of bug:
https://bugs.winehq.org/show_bug.cgi?id=30814

An analysis of the bug can be found:
https://davidgow.net/hacks/aoescroll.html

Fixes: d2cbafab08 ("user32: Use input shared memory in GetKeyboardState.")
Signed-off-by: David Gow <david@ingeniumdigital.com>
---
 dlls/win32u/input.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/dlls/win32u/input.c b/dlls/win32u/input.c
index 90aa6754313..79112b8b76d 100644
--- a/dlls/win32u/input.c
+++ b/dlls/win32u/input.c
@@ -167,7 +167,11 @@ BOOL WINAPI NtUserGetKeyboardState( BYTE *state )
     }
     SHARED_READ_END( &shared->seq );
 
-    if (skip) return TRUE;
+    if (skip)
+    {
+        for (i = 0; i < 256; i++) state[i] &= 0x81;
+        return TRUE;
+    }
 
     memset( state, 0, 256 );
     SERVER_START_REQ( get_key_state )
